lib/Conversion/TorchToLinalg/TorchToLinalg.cpp

```diff
@@ -1720,14 +1720,25 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
   }
 
   if (auto gtScalar = dyn_cast<AtenGtScalarOp>(op)) {
-    Type dtype = gtScalar.self().getType().cast<ValueTensorType>().getDtype();
-    if (!dtype.isa<mlir::FloatType>()) {
-      gtScalar.emitError("unimplemented: non-floating point operand dtype");
-      return nullptr;
+    Type dtype = gtScalar.self().getType().cast<BaseTensorType>().getDtype();
+
+    // TODO: `gtTensor` and `gtScalar` share similar code and can be called from
```

Comment:
One side comment thought: Would it make sense to add a rule to ReduceOpVariants to convert `aten.foo.Scalar(x, y)` to `aten.foo.Tensor(x, NumToTensor(y))`? I think we can detect this case and add a trait "ScalarVariantWithCorrespondingTensorVariant" to the ops that have this property. It would look similar to how we handle emit_with_mutating_variants in https://github.com/llvm/torch-mlir/blob/528354de84c5892d539fce17c74b2c928a645991/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py#L430

We would probably rename emit_with_mutating_variants -> emit_elementwise_op and have it detect which mutating/scalar/tensor variants exist, emit them, and add all the necessary traits.


Will let Ramiro drive this review, which seems to be going in a good direction anyway. Please submit after he gives approval.

Reply:
I think adding something like this would be really nice. We just have to be careful about the type conversions that happened when a scalar is turned into a tensor. In particular, `NumToTensor` would have the wrong behavior because, for example, it converts scalar floats to `float64`, which would mess with the type promotion when performing the tensor computation. What we want is something that behaves like the `wrapped_number` in PyTorch, which follows a different set of type conversion rules.

Reply:
+1, yeah, there are probably some subtle cases.